File: 2026/aed-1/cormen/12.1/Main.java
public class Main {
    public static void main(String[] args) {
        Tree<Integer> tree = new Tree<>(6, 6);
        Node<Integer> sevenNode = new Node<Integer>(7, 7);

        tree.insert(new Node<Integer>(5, 5));
        tree.insert(new Node<Integer>(2, 2));
        tree.insert(new Node<Integer>(5, 5));
        tree.insert(sevenNode);
        tree.insert(new Node<Integer>(8, 8));

        Node<Integer> found = tree.search(tree.root, 8);

        if (found != null) {
            IO.println("found: " + found.data);
        }

        Node<Integer> minimum = tree.minimum(tree.root);

        if (minimum != null) {
            IO.println("minimum: " + minimum.data);
        }

        Node<Integer> maximum = tree.maximum(tree.root);

        if (maximum != null) {
            IO.println("maximum: " + maximum.data);
        }

        Node<Integer> predecessor = tree.predecessor(sevenNode);

        if (predecessor != null) {
            IO.println("predecessor of 7: " + predecessor.data);
        }

        Node<Integer> successor = tree.successor(sevenNode);

        if (successor != null) {
            IO.println("successor of 7: " + successor.data);
        }

        IO.println("preorder walk:");

        tree.preorderWalk(tree.root);

        IO.println("inorder wak:");

        tree.inorderWalk(tree.root);

        IO.println();

        tree.inorderWalkNonRecursive(tree.root);

        IO.println("postorder walk:");

        tree.postorderWalk(tree.root);

        IO.println("inorder walk after deleting 7:");

        tree.delete(sevenNode);

        tree.inorderWalk(tree.root);
    }
}

File: 2026/aed-1/cormen/12.1/main.py
from Node import Node
from Tree import Tree

def main() -> None:
    tree: Tree = Tree(6, 6)
    sevenNode: Node = Node(7, 7)

    tree.insert(Node(5, 5))
    tree.insert(Node(2, 2))
    tree.insert(Node(5, 5))
    tree.insert(sevenNode)
    tree.insert(Node(8, 8))

    found: Node | None = tree.search(tree.root, 8)

    if found:
        print(f"found: {found.data}")

    minimum: Node | None = tree.minimum(tree.root)

    if minimum:
        print(f"minimum: {minimum.data}")

    maximum: Node | None = tree.maximum(tree.root)

    if maximum:
        print(f"maximum: {maximum.data}")

    predecessor: Node | None = tree.predecessor(sevenNode)

    if predecessor:
        print(f"predecessor of 7: {predecessor.data}")

    successor: Node | None = tree.successor(sevenNode)

    if successor:
        print(f"successor of 7: {successor.data}")

    print("preorder walk:")

    tree.preorder_walk(tree.root)

    print("inorder wak:")

    tree.inorder_walk(tree.root)

    print("")

    tree.inorder_walk_non_recursive(tree.root)

    print("postorder walk:")

    tree.postorder_walk(tree.root)

    print("inorder walk after deleting 7:")

    tree.delete(sevenNode)

    tree.inorder_walk(tree.root)


if __name__ == '__main__':
    main()
